### DISCUSSION TITLE
Raising an error if input model has multiple valid parameters?

### DISCUSSION THREAD

**Discussion by rgstroh** [fastapi/fastapi#6417] [Questions] [answered]
### Description

This is similar to the question https://github.com/tiangolo/fastapi/issues/724#issue-526375835 but instead of extra parameters, how do I raise an error if multiple of the same valid parameters are provided.

**Example:**
```
from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()

class UpdateModel(BaseModel):
    name: str
    class Config:
        extra = "forbid"

@app.post("/items/")
async def update(item: UpdateModel):
    # do smth with item.name
```

If I send {"name": "FOO", "name": "BAR"} the second "name" is used for item (i.e. item.name = "BAR")
How can I raise an error with my pydantic model so this is rejected?  The extra = "forbid" does not work for duplicates.

Thanks

### Operating System

Windows

### FastAPI Version

0.66.0

### Python Version

3.9.1

**Comment by Dustyposa** [comment]
Sorry, how can you send json data with two same key? 
If you want to get `name ` and `item name` in json body.  You can do these ways:
- direct send 
    ```json
    {"name": "FOO", "item_name": "BAR"} 
    ```
- extra model send:
   ```json
   {"name": "FOO", "item": {"name": "BAR"}} 
   ```

**Comment by rgstroh** [comment]
Maybe unlikely, but if client app has an error and sends a json body with {"name": "FOO1", "name": "FOO2"} I would want pydantic to raise an error.  This would make the API more robust.

**Comment by insomnes** [comment]
I believe that it is result of python's json decoding. You can us some `object_pairs_hook` function in [JSONDecoder](https://docs.python.org/3/library/json.html#json.JSONDecoder) creation, which will raise error if key duplication found.
You can change JSON deserialization function in pydantic's [model config](https://pydantic-docs.helpmanual.io/usage/exporting_models/#custom-json-deserialisation).
![image](https://user-images.githubusercontent.com/16958893/137360879-66b28f8c-06c5-469f-89f9-3d16c3e9783b.png)

**Comment by rgstroh** [comment]
I tried implementing the below code so the Config class uses a custom decoder, but it doesn't appear to be called.  What's interesting is when I add a custom encoder for "json_dumps" this does get called, but the dictionary coming into this function already has the duplicated parameter filtered out.

```
def raise_error_if_dup_key(pairs):
    result = {}
    for key, value in pairs:
        if key in result:
            raise ValueError(f"Duplicate key found:{key}")
        result[key] = value
    return result

def custom_loads(s):
    return json.loads(s,object_pairs_hook = raise_error_if_dup_key)

class UpdateModel(BaseModel):
    name: str
    class Config:
        extra = "forbid"
        json_loads = custom_loads
```

**Comment by insomnes** [comment, marked as answer]
That was interesting ride for me.
FastAPI decodes JSON from starlette's Request.json().
So to change JSON decoding we need to add custom Request and APIRoute (as mentioned in [docs](https://fastapi.tiangolo.com/advanced/custom-request-and-route/)):
```python3
import json
from typing import Any, Callable, Coroutine


from fastapi import FastAPI, Request, Response
from fastapi.routing import APIRoute
from pydantic import BaseModel


def raise_error_if_dup_keys(pairs):
    result = {}
    for key, value in pairs:
        if key in result:
            raise ValueError(f"Duplicate key found in json: {key}")
        result[key] = value
    return result


class NoDupsJSONRequest(Request):
    async def json(self) -> Any:
        if not hasattr(self, "_json"):
            body = await self.body()
            self._json = json.loads(body, object_pairs_hook=raise_error_if_dup_keys)
        return self._json


class NoDupsRoute(APIRoute):
    def get_route_handler(self) -> Callable[[Request], Coroutine[Any, Any, Response]]:
        original_route_handler = super().get_route_handler()

        async def custom_route_handler(request: Request) -> Response:
            request = NoDupsJSONRequest(request.scope, request.receive)
            return await original_route_handler(request)

        return custom_route_handler


class UpdateModel(BaseModel):
    name: str


app = FastAPI()
app.router.route_class = NoDupsRoute


@app.post("/items/")
async def update(item: UpdateModel):
    return item
```
This will result in:
```
curl -X POST  http://127.0.0.1:8000/items/ -H 'Content-Type: application/json' -d '{"name": "foo", "name": "bar"}'

{"detail":"There was an error parsing the body"}%                                                                                                                                                             
```

You can get more verbose message if you raise `json.JSONDecodeError`, but i don't want to lurk more in it's parameters.

**Comment by rgstroh** [comment]
wow, thank you for digging into this!
